src/software/backend/simulation/simulator_robot.cpp

```diff
@@ -1,382 +1,437 @@
+        return robot->getMotorSpeedFrontRight();
     }
+    LOG(WARNING) << "SimulatorRobot being used with invalid PhysicsRobot" << std::endl;
     return 0.0;
 }
 
-void SimulatorRobotSingleton::coastMotorBackLeft()
+void SimulatorRobot::coastMotorFrontLeft()
 {
-    if (auto robot = getCurrentPhysicsRobot().lock())
-    {
-        // TODO: Implement me
-    }
+    // We coast by simply doing nothing and not applying wheel force
 }
 
-void SimulatorRobotSingleton::coastMotorBackRight()
+void SimulatorRobot::coastMotorBackLeft()
 {
-    if (auto robot = getCurrentPhysicsRobot().lock())
-    {
-        // TODO: Implement me
-    }
+    // We coast by simply doing nothing and not applying wheel force
 }
 
-void SimulatorRobotSingleton::coastMotorFrontLeft()
+void SimulatorRobot::coastMotorBackRight()
 {
-    if (auto robot = getCurrentPhysicsRobot().lock())
-    {
-        // TODO: Implement me
-    }
+    // We coast by simply doing nothing and not applying wheel force
 }
 
-void SimulatorRobotSingleton::coastMotorFrontRight()
+void SimulatorRobot::coastMotorFrontRight()
 {
-    if (auto robot = getCurrentPhysicsRobot().lock())
-    {
-        // TODO: Implement me
-    }
+    // We coast by simply doing nothing and not applying wheel force
 }
 
-void SimulatorRobotSingleton::brakeMotorBackLeft()
+void SimulatorRobot::brakeMotorFrontLeft()
 {
-    if (auto robot = getCurrentPhysicsRobot().lock())
+    if (auto robot = physics_robot.lock())
     {
-        // TODO: Implement me
+        robot->brakeMotorFrontLeft();
+    }
+    else
+    {
+        LOG(WARNING) << "SimulatorRobot being used with invalid PhysicsRobot"
+                     << std::endl;
     }
 }
 
-void SimulatorRobotSingleton::brakeMotorBackRight()
+void SimulatorRobot::brakeMotorBackLeft()
 {
-    if (auto robot = getCurrentPhysicsRobot().lock())
+    if (auto robot = physics_robot.lock())
+    {
+        robot->brakeMotorBackLeft();
+    }
+    else
     {
-        // TODO: Implement me
+        LOG(WARNING) << "SimulatorRobot being used with invalid PhysicsRobot"
+                     << std::endl;
     }
 }
 
-void SimulatorRobotSingleton::brakeMotorFrontLeft()
+void SimulatorRobot::brakeMotorBackRight()
 {
-    if (auto robot = getCurrentPhysicsRobot().lock())
+    if (auto robot = physics_robot.lock())
+    {
+        robot->brakeMotorBackRight();
+    }
+    else
     {
-        // TODO: Implement me
+        LOG(WARNING) << "SimulatorRobot being used with invalid PhysicsRobot"
+                     << std::endl;
     }
 }
 
-void SimulatorRobotSingleton::brakeMotorFrontRight()
+void SimulatorRobot::brakeMotorFrontRight()
 {
-    if (auto robot = getCurrentPhysicsRobot().lock())
+    if (auto robot = physics_robot.lock())
+    {
+        robot->brakeMotorFrontRight();
+    }
+    else
     {
-        // TODO: Implement me
+        LOG(WARNING) << "SimulatorRobot being used with invalid PhysicsRobot"
+                     << std::endl;
     }
```

Comment:
This if-else check is being used a _ton_. Perhaps we could put it in a macro? very roughly: `DO_A_THING_AND_WARN_IF_INVALID_PHYSICS_ROBOT(robot->brakeMotorFrontRight())`

Maybe we can use this pattern elsewhere as well?

Reply:
Yea I thought of something like that but just wasn't sure if I should or not, but that's likely a good idea since it is used a lot.

What lambda functions perhaps be better, especially if we ever needed slightly more complex logic?

ie.
```
callFuncAndWarnIfInvalid([this](){robot->brakeMotorFrontRight();});
```

Would be slightly more verbose to setup the lambda but handle complexity better

Reply:
Yeah, verbosity is the reason I didn't immediately go for lambdas, but this doesn't look too bad, so I'd say let's go for lambdas. This should help chop out a lot of lines :smile:

Reply:
changed to use lambdas in `simulator_ball`, `simulator_ball_singleton`, `simulator_robot`, `simulator_robot_singleton`